README.md
# tmap: thematic maps in R <img src="man/figures/logo.png" align="right" height="139" alt="" />

<!-- badges: start -->
[![R CMD CHECK](https://github.com/r-tmap/tmap/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/r-tmap/tmap/actions/workflows/R-CMD-check.yaml)
[![Codecov test coverage](https://codecov.io/gh/r-tmap/tmap/branch/master/graph/badge.svg)](https://app.codecov.io/gh/r-tmap/tmap?branch=master)
[![CRAN](http://www.r-pkg.org/badges/version/tmap)](https://cran.r-project.org/package=tmap) 
[![CRAN checks](https://cranchecks.info/badges/worst/tmap)](https://cran.r-project.org/web/checks/check_results_tmap.html)
[![Downloads](http://cranlogs.r-pkg.org/badges/tmap?color=brightgreen)](https://www.r-pkg.org:443/pkg/tmap)
[![License](https://img.shields.io/badge/License-GPL%20v3-brightgreen.svg?style=flat)](http://www.gnu.org/licenses/gpl-3.0.html)
[![r-universe](https://r-tmap.r-universe.dev/badges/tmap)](https://r-tmap.r-universe.dev/tmap)

<!-- badges: end -->



[`tmap`][1] is an actively maintained open-source [R][2]-library for drawing thematic maps. The API is based on [*A Layered Grammar of Graphics*][4] and resembles the syntax of [`ggplot2`][3], a popular R-library for drawing charts.


<span>
<img src="https://mtennekes.github.io/downloads/images/classic.png" alt="Classic World Map" height="125px"/>
<img src="https://mtennekes.github.io/downloads/images/bubble.png" alt="Bubble Map" height="125px"/>
<img src="https://mtennekes.github.io/downloads/images/view_metro4.jpg" alt="Bubble Map" height="125px"/>
<img src="https://mtennekes.github.io/downloads/images/world_facets2.png" alt="World facets" height="125px"/>
<img src="https://mtennekes.github.io/downloads/images/USchoro.png" alt="US Choropleth" height="125px"/>
<img src="https://mtennekes.github.io/downloads/images/US_PR.jpg" alt="US Choropleth" height="125px"/>
<img src="https://mtennekes.github.io/downloads/images/crimes3b.png" alt="London Crimes" height="125px"/>
<img src="https://mtennekes.github.io/downloads/images/crimes4.png" alt="London Crimes" height="125px"/>
<img src="https://mtennekes.github.io/downloads/images/crimes5.png" alt="London Crimes" height="125px"/>
<img src="https://mtennekes.github.io/downloads/images/view_crimes2.JPG" alt="London Crimes" height="125px"/>
</span>

See [below](https://github.com/r-tmap/tmap/#reference) the source code for these images.


Installation
------------

`tmap` is available on [![CRAN](http://www.r-pkg.org/badges/version/tmap)](https://cran.r-project.org/package=tmap). Installation is straightforward:

```r
install.packages("tmap")
```

For Linux and macOS users who are new to working with spatial data in R, this may fail since additional (non-R) libraries are required (which are automatically installed for Windows users).

**Windows**
No additional installation required.

**Linux (Ubuntu)**
See https://geocompx.org/post/2020/installing-r-spatial-packages-linux/. Please address installation issues in this [issue](https://github.com/r-tmap/tmap/issues/150).

**macOS**
See http://www.kyngchaos.com. Please address installation issues in this [issue](https://github.com/r-tmap/tmap/issues/149).


Development
------------


The latest development version can be installed using `remotes`. 
In order to use the development version of `tmap`, it is recommended to use the development version of [tmaptools](https://github.com/r-tmap/tmaptools) as well:

```r
# install.packages("remotes")
install_github("r-tmap/tmaptools")
install_github("r-tmap/tmap")

# On Linux, with pak
# install.packages("pak")
pak::pak("r-tmap/tmap")

# Or from r-universe
install.packages("tmap", repos = c("https://r-tmap.r-universe.dev", "https://cloud.r-project.org"))
```

See [NEWS](https://github.com/r-tmap/tmap/blob/master/NEWS.md) for the latest features and improvements and the [issue list](https://github.com/r-tmap/tmap/issues) for discussions of enhancements and bugs.

Do you want to help with the development of the package?
Please let us know! 
Any feedback, requests, tips, or bug reports are welcome!

Usage
-----

```r
library(tmap)
```

Reference
----
[Tennekes, M., 2018, tmap: Thematic Maps in R, Journal of Statistical Software, 84(6), 1-39](https://doi.org/10.18637/jss.v084.i06)

This paper has been written for `tmap` 1.11-2. See the [reproducible code](https://cran.r-project.org/package=tmap/vignettes/tmap-JSS-code.html) using `tmap` 2.x.

Vignettes 
-----

[tmap v3: get started!](https://cran.r-project.org/package=tmap/vignettes/tmap-getstarted.html)

[tmap: what is changed in recent versions?](https://cran.r-project.org/package=tmap/vignettes/tmap-changes.html)


Tutorials
-----

[tmap in RMarkdown](https://github.com/r-tmap/tmap/blob/master/demo/tutorials/rmarkdown_tmap.Rmd)


Presentations
-----

* [Creating thematic maps in R][23] (OpenGeoHUB Summer School 2019)
* [Plotting spatial data with R](https://github.com/mtennekes/tmap-workshop) (eRum 2018) 
* [Exploring and presenting maps with tmap][21] (useR!2017)
* [tmap: creating thematic maps in a flexible way][10] (useR!2015)]


Other resources
-----

* [Geocomputation with R; Making maps with R, Robin Lovelace, Jakub Nowosad, Jannes Muenchow][20]
* [R tip: Create maps in R][24]
* [Working with Spatial Data and using tmap, Samantha A. Alger][22]
* [Computer World: Mapping in R just got a whole lot easier][18]
* [National Socio-Environmental Synthesis Center: Maps in R][19]
* [Introduction to visualizing spatial data in R][9]
* [Blog post SpatialControl][7]
* [Blog post TWIAV][8]
* [Computer World: Create maps in R in 10 (fairly) easy steps][12]
* [Computer World: Great R packages for data import, wrangling and visualization][17]
* [Tutorial Visualizing spatial data: from base to shiny - workshop][15]
* [Stack Overflow questions (#tmap)][16]


Getting help
-----

There are two main places to get help with `tmap`:

1.  [stackoverflow](https://stackoverflow.com/tags/tmap) is a great source of answers to common tmap questions. 
It is also a great place to get help, once you have created a reproducible example that illustrates your problem. 
Please tag your questions with *tmap*.
2.  If you have a request or if think your problem is caused by a bug, please open an [issue](https://github.com/r-tmap/tmap/issues), preferably with a reproducible example.


  [1]: https://cran.r-project.org/package=tmap
  [2]: https://stackoverflow.com/tags/r/info
  [3]: https://cran.r-project.org/package=ggplot2
  [4]: http://vita.had.co.nz/papers/layered-grammar.pdf
  [5]: https://github.com/r-tmap/tmap
  [6]: https://cran.r-project.org/web/packages/tmap/vignettes/tmap-nutshell.html
  [7]: http://spatcontrol.net/SpatialControl/2015/11/06/tmap-r-package/
  [8]: https://twiav.nl/en/blog0002en.php
  [9]: https://cran.r-project.org/doc/contrib/intro-spatial-rl.pdf
  [10]: https://mtennekes.github.io/downloads/presentations/tmap_user2015.pdf
  [11]: https://cran.r-project.org/web/packages/tmap/vignettes/tmap-modes.html
  [12]: http://cwrld.us/Rmaps10
  [13]: https://github.com/r-tmap/tmap/blob/master/demo/US_choropleth.R
  [14]: https://github.com/r-tmap/tmap/blob/master/demo/crimes_in_Greater_London.R
  [15]: https://github.com/Robinlovelace/Creating-maps-in-R/blob/master/vignettes/vspd-base-shiny.Rmd
  [16]: https://stackoverflow.com/questions/tagged/tmap
  [17]: http://www.computerworld.com/article/2921176/business-intelligence/great-r-packages-for-data-import-wrangling-visualization.html
  [18]:	http://www.computerworld.com/article/3175623/data-analytics/mapping-in-r-just-got-a-whole-lot-easier.html
  [19]: https://sesync-ci.github.io/maps-in-R-lesson/
  [20]: https://r.geocompx.org/adv-map.html
  [21]: https://mtennekes.github.io/downloads/presentations/tmap_user2017.pdf
  [22]: https://gotellilab.github.io/Bio381/StudentPresentations/SpatialDataTutorial.html
  [23]: https://mtennekes.github.io/downloads/presentations/tmap_opengeo_muenster.pdf
  [24]: https://www.youtube.com/watch?v=wgFVmzSbaQc#t=3m20s
  


NEWS
version 3.3-4
- (!) last version of tmap 3.x. Next CRAN version will be tmap 4.x (release planned at the end of 2023)
- fixed bug (some stars appeared upside down in plot mode)
- fixed newly introduced shiny bug (#767)

version 3.3-3
- added device to tmap_save
- fixed many small bugs (see github issue list)

version 3.3-2
- World has been fixed (it is now 's2' proof, see github issue #564)

version 3.3-1
- fixed view titles
- added html.escape to popup.format to control whether html code is escaped in view popups
- fixed many small bugs (see github issue list)

version 3.3
- tmap_grob added, which exports tmap plots to grob objects (grid package)
- SpatRaster objects (terra package) supported
- fixed many small bugs (see github issue list)

version 3.2
- tmap_animation improved: now using av and gifski under the hood; added progress bar; lists of tmap objects supported; improved default settings
- improved tmapOutput; it now works with reactive shiny objects
- improved internally used margins; also small exported maps look good (see example tmap_save)
- improved tmap_design_mode; tmap.design.mode is now a global option (and not a tmap option anymore)
- reexported providers from leafet
- added show.warnings to tmap_options
- added "width" argument to tm_credits
- tm_text improved: added "id" argument, and tm_remove_layer can be applied to it

version 3.1
- interactive maps in origin CRS working: tmap_options(projection = 0, basemaps = NULL)
- added tm_mouse_coordinates() to show mouse coordinates in view mode
- added tmap_design_mode() to toggle the desgin mode.
- made backgrond symbol grob shapes transparent
- added in.iframe and selfcontained to tmap_save
- improved tm_add_legend: added type = "title" for title only legend elements
- added ttmp() which shows the last map in the other mode. 
- fixed many small bugs (see github issue list)

version 3.0
- [!] spatial rasters/arrays now implemented using the stars package (instead of the raster package)
- as.count argument added to layer functions; numeric variables can be treated as count variables (integers)
- drop.levels argument added to layer functions which allows to drop unused levels
- new style added for color classes: "dpih" and "headtails"; also added style.args to pass on arguments to classIntervals
- grid labels specification per axis
- fixed bug with geometry collection

version 2.3-2
- probably last version before release of tmap 3.0
- many small improvements and bug fixes (see github)
- first stars supported (full support expected in tmap 3.0)
- replaced projection shortcuts with epsg/proj4strings in examples

version 2.3-1
- improved popups in view mode
- added validity checks for sf objects
- minor bugs fixed: see github issue list

version 2.3
- [!] shiny integration: added tmapOutput, renderTmap, tmapProxy, and, tm_remove_layer
- improved tmap_save (regarding default values)
- improved tm_rgb (added r g b and max.value arguments)
- added tm_graticules
- supported TMS servers
- in view mode, layers can be hidden from the legend with group = NULL
- replaced mapview by leafsync
- minor bugs fixed: see github issue list

version 2.2
- improved tm_rgb
- improved breaks in interactive maps
- added bbox argument to qtm
- improved projection shortcut codes (using EPSG database)
- fixed several bugs (see github)

version 2.1-1
- fixed two small bugs

version 2.1
- added option max.raster (maximum raster size)
- made rasters in view mode faster
- bug fixed for geometry collection objects without data in view mode
- improved handling of manual palettes
- tmap_arrange can take a list of tmap objects

version 2.0
- [!] tmap migrated from sp to sf. sp objects are still supported, but all internal functions are based on sf
- [!] added tm_sf which draws sf objects
- [!] renamed all xxx_tmap functions (e.g. save_tmap) to tmap_xxx for consistency
- [!] added tm_basemap and tm_tiles: now, multiple (overlay) tiled maps can be shown
- [!] rearranged tmap options, styles and formats. See tmap_options, tmap_style, tmap_format, tm_style and tm_format
- [!] added filter argument to tm_shape and added colorNULL to layer functions
- added tm_minimap
- added remove.NA.facets to tm_facets
- units data columns are supported
- tmap_animation enhanced with loop options
- added tmap_tips()
- all changes are backwards compatible with 1.11-2
- vignettes rewritten, including a vignette that describes the changes of 2.0

version 1.11-2
- added references to JSS paper (including citation)

version 1.11-1
- fixed bug in labels argument of tm_fill
- fixed bug regarding legend symbols in facets with free scales
- fixed bug in rasters in view mode
- improved popup width in view mode

version 1.11
- added text.align and text.to.columns to legend.format
- legend.(aes).reverse added to the layer functions
- symbol shapes can be put in a list for small multiples, and named according the values of the variable specified with the shape argument
- grid labels can now be rotated and formatted
- changed default values of free.coords and drop.units in tm_facets
- fixed bug with all NA's in view mode
- fixed bug with dismo::gmap raster shapes
- fixed bug for ambiguous values for aesthetics (e.g. when "blue" is also a variable name)
- NOTE: this will be the last version before the major update (in which sf fully replaces sp)

version 1.10
- [!] tm_scale_bar, tm_text, and tm_grid now supported in view mode (required leaflet 1.1.0)
- added along argument to tm_facets, which enables facetting along multiple pages (or animation frames)
- added main.title argument to tm_layout
- added tmap_options, including the new option tmap.unit ("metric" or "imperial"")
- improved automatic labeling of breaks
- improved legend in view mode: missing values are shown by default (like in plot mode)
- improved bubble scaling in view mode
- fixed bug aspect ratio bug

version 1.8-1
- added tm_rgb (shortcut for rgb rasters)
- fixed interactive maps in rmarkdown

R/qtm.R
#' Quick thematic map plot
#' 
#' Draw a thematic map quickly. This function is a convenient wrapper of the main
#' plotting method of stacking [`tmap-element`]s. Without arguments or with a
#' search term, this functions draws an interactive map.
#' 
#' The first argument is a shape object (normally specified by [tm_shape()]).
#' The next arguments, from `fill` to `raster`, are the aesthetics from the main
#' layers. The remaining arguments are related to the map layout. Any argument
#' from any main layer function, such as [tm_polygons()], can be specified (see `...`).
#' It is also possible to stack [`tmap-element`]s on a `qtm` plot. See examples.
#' 
#' By default, a scale bar is shown. This option can be set with [tmap_options()] 
#' (argument `qtm.scalebar`). A minimap is shown by default when `qtm` is called
#' without arguments of with a search term. This option can be set with [tmap_options()]
#' (argument `qtm.minimap`).
#' 
#' @param shp One of:
#'  * shape object, which is an object from a class defined by the [`sf`][`sf::sf`]
#'    or [`stars`][stars::st_as_stars()] package. Objects from the packages `sp`
#'    and `raster` are also supported, but discouraged.
#'  * Not specified, i.e. `qtm()` is executed. In this case a plain interactive
#'    map is shown.
#'  * An OpenStreetMap search string, e.g. `qtm("Amsterdam")`. In this case a
#'    plain interactive map is shown positioned according to the results of the
#'    search query (from OpenStreetMap nominatim)
#' @param fill,col,size,shape,lwd,lty,fill_alpha,col_alpha Visual variables.
#' @param by data variable name by which the data is split, or a vector of two
#'   variable names to split the data by two variables (where the first is used
#'   for the rows and the second for the columns). See also [tm_facets()].
#' @param scale numeric value that serves as the global scale parameter. All font
#'   sizes, symbol sizes, border widths, and line widths are controlled by this value.
#'   The parameters `symbols.size`, `text.size`, and `lines.lwd` can be scaled
#'   separately with respectively `symbols.scale`, `text.scale`, and 
#'   `lines.scale`. See also `...`.
#' @param title main title. For legend titles, use `X.style`, where X is the
#'   layer name (see `...`).
#' @param crs Either a [`crs`][sf::st_crs()] object or a character value
#'   (`PROJ.4` character string). By default, the projection is used that is
#'   defined in the `shp` object itself.
#' @param bbox bounding box. Argument passed on to [tm_shape()]
#' @param basemaps name(s) of the provider or an URL of a tiled basemap.
#'   It is a shortcut to [tm_basemap()]. Set to `NULL` to disable basemaps.
#'   By default, it is set to the tmap option `basemaps`.
#' @param overlays name(s) of the provider or an URL of a tiled overlay map.
#'   It is a shortcut to [tm_tiles()].
#' @param zindex zindex
#' @param group group
#' @param group.control group.control
#' @param style Layout options (see [tm_layout()]) that define the style.
#'   See [tmap_style()] for details.
#' @param format Layout options (see [tm_layout()]) that define the format.
#'   See [tmap_format()] for details.
#' @param ... arguments passed on to the `tm_*` functions. The prefix of these
#'   arguments should be with the layer function name without `"tm_"` and a period.
#'   For instance, the palette for polygon fill color is called `fill.palette`.
#'   The following prefixes are supported: `shape.`, `fill.`, `borders.`, `polygons.`,
#'    `symbols.`, `dots.`, `lines.`, `raster.`, `text.`, `layout.`, `grid.`,
#'     `facets.`, and `view.`. Arguments that have a unique name, i.e. that does
#'     not exist in any other layer function, e.g. `convert2density`, can also be
#'     called without prefix.
#' @return A [`tmap-element`]
#' @example ./examples/qtm.R
#' @seealso `vignette("tmap_sneak_peek")`
#' @references Tennekes, M., 2018, {tmap}: Thematic Maps in {R},
#'   Journal of Statistical Software, 84(6), 1-39, \doi{10.18637/jss.v084.i06}
#' @export
qtm <- function(shp, 
				fill = tm_const(),
				col = tm_const(),
				size = tm_const(),
				shape = tm_const(),
				lwd = tm_const(),
				lty = tm_const(),
				fill_alpha = tm_const(),
				col_alpha = tm_const(),
				by = NULL,
				scale = NULL,
				title = NULL,
				crs = NULL,
				bbox = NULL,
				basemaps = NULL,
				overlays = NULL,
				zindex = NA,
				group = NA,
				group.control = "check",
				style=NULL,
				format=NULL,
				...) {
	args <- list(...)
	shp_name <- deparse(substitute(shp))[1]
	called <- names(match.call(expand.dots = TRUE)[-1])
	
	tmapOptions <- tmap_options_mode()
	show.warnings = tmapOptions$show.warnings
	
	if (missing(shp) || is.character(shp)) {
		
		viewargs <- args[intersect(names(args), names(formals(tm_view)))]
		
		if (!missing(shp)) viewargs$bbox <- shp
		
		g <- c(tm_basemap(basemaps), tm_tiles(overlays), do.call("tm_view", viewargs))
		attr(g, "qtm_shortcut") <- TRUE
		class(g) <- "tmap"
		return(g)
	}

	
	lst = c(list(size = size, 
				 fill = fill, 
				 col = col,
				 shape = shape,
				 lwd = lwd,
				 lty = lty,
				 fill_alpha = fill_alpha,
				 col_alpha = col_alpha,
				 zindex = zindex,
				 group = group,
				 group.control = group.control), args)
	
	# if shape is specified at tm_sf, symbols are drawn instead of dots
	if (!"shape" %in% called) lst$shape = NULL
	
	g = tm_shape(shp) + do.call(tm_sf, lst)
	
	assign("last_map_new", match.call(), envir = .TMAP)
	attr(g, "qtm_shortcut") <- FALSE
	g
}
